confidant/authnz/userauth.py

```diff
@@ -0,0 +1,661 @@
+        root_url = root_url.rstrip('/')
+
+        # TODO: also support unspecified?
+        name_id_fmt = 'urn:oasis:names:tc:SAML:1.1:nameid-format:emailAddress'
+
+        # Service Provider section
+        sp_data = {
+            'entityId': root_url + '/v1/saml/metadata',
+            'assertionConsumerService': {
+                'url': root_url + '/v1/saml/consume',
+                'binding': 'urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST'
+            },
+            'singleLogoutService': {
+                'url': root_url + '/v1/saml/logout',
+                'binding': 'urn:oasis:names:tc:SAML:2.0:bindings:HTTP-REDIRECT'
+            },
+            'NameIDFormat': name_id_fmt,
+        }
+
+        sp_has_key = False
+        if app.config['SAML_SP_KEY_FILE']:
+            sp_has_key = True
+            sp_data['privateKey'] = self._load_rsa_for_saml(
+                app.config['SAML_SP_KEY_FILE'],
+                password=app.config.get('SAML_SP_KEY_FILE_PASSWORD'))
+        if app.config['SAML_SP_KEY']:
+            sp_has_key = True
+            sp_data['privateKey'] = app.config['SAML_SP_KEY']
+
+        if app.config['SAML_SP_CERT_FILE']:
+            sp_data['x509cert'] = self._load_x509_for_saml(
+                app.config['SAML_SP_CERT_FILE'])
+        if app.config['SAML_SP_CERT']:
+            sp_data['x509cert'] = app.config['SAML_SP_CERT']
+
+        # security defaults: sign everything if SP key was provided
+        security_data = {
+            'nameIdEncrypted': False,
+            'authnRequestsSigned': sp_has_key,
+            'logoutRequestsSigned': sp_has_key,
+            'logoutResponsesSigned':
+                app.config['SAML_SECURITY_SLO_RESP_SIGNED'],
+            'signMetadata': sp_has_key,
+            'wantMessagesSigned':
+                app.config['SAML_SECURITY_MESSAGES_SIGNED'],
+            'wantAssertionsSigned':
+                app.config['SAML_SECURITY_ASSERTIONS_SIGNED'],
+            'wantNameIdEncrypted': False,
+            "signatureAlgorithm": app.config['SAML_SECURITY_SIG_ALGO'],
+        }
+
+        # Identity provider section
+        idp_data = {
+            'entityId': app.config['SAML_IDP_ENTITY_ID'],
+            'singleSignOnService': {
+                'url': app.config['SAML_IDP_SIGNON_URL'],
+                'binding': 'urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect'
+            },
+        }
+
+        if app.config['SAML_IDP_LOGOUT_URL']:
+            idp_data['singleLogoutService'] = {
+                'url': app.config['SAML_IDP_LOGOUT_URL'],
+                'binding': 'urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect'
+            }
+
+        if app.config['SAML_IDP_CERT_FILE']:
+            idp_data['x509cert'] = self._load_x509_for_saml(
+                app.config['SAML_IDP_CERT_FILE'])
+        if app.config['SAML_IDP_CERT']:
+            idp_data['x509cert'] = app.config['SAML_IDP_CERT']
```

Comment:
Similar story to the above SP_CERT stuff. IDP messages and assertions may optionally be signed or encrypted or both.

If you've set the `SAML_SECURITY_MESSAGES_SIGNED` option, then python-saml will at runtime reject IDP messages that are not signed, or it will raise a runtime exception if there is no IDP certificate. I didn't add additional validation that the settings seem self-consistent here because it seemed like second guessing validation that only exists in python-saml at runtime right now.